community/index/src/test/java/org/neo4j/index/internal/gbptree/GBPTreeLockTest.java

```diff
@@ -0,0 +1,189 @@
+ */
+package org.neo4j.index.internal.gbptree;
+
+import org.junit.Test;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.junit.Assert.fail;
+
+public class GBPTreeLockTest
+{
+    // Lock can be in following states and this test verify transitions back and forth between states
+    // and also verify expected behaviour after each transition.
+    //            Writer   | Cleaner
+    // State UU - unlocked | unlocked
+    // State UL - unlocked | locked
+    // State LU - locked   | unlocked
+    // State LL - locked   | locked
+
+    private GBPTreeLock lock = new GBPTreeLock();
+    private GBPTreeLock copy;
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Test
+    public void test_UU_UL_UU() throws Exception
+    {
+        // given
+        assertUU();
+
+        // then
+        lock.cleanerLock();
+        assertUL();
+
+        lock.cleanerUnlock();
+        assertUU();
+    }
+
+    @Test
+    public void test_UL_LL_UL() throws Exception
+    {
+        // given
+        lock.cleanerLock();
+        assertUL();
+
+        // then
+        lock.writerLock();
+        assertLL();
+
+        lock.writerUnlock();
+        assertUL();
+    }
+
+    @Test
+    public void test_LL_LU_LL() throws Exception
+    {
+        // given
+        lock.writerLock();
+        lock.cleanerLock();
+        assertLL();
+
+        // then
+        lock.cleanerUnlock();
+        assertLU();
+
+        lock.cleanerLock();
+        assertLL();
+    }
+
+    @Test
+    public void test_LU_UU_LU() throws Exception
+    {
+        // given
+        lock.writerLock();
+        assertLU();
+
+        // then
+        lock.writerUnlock();
+        assertUU();
+
+        lock.writerLock();
+        assertLU();
+    }
+
+    @Test
+    public void test_UU_LL_UU() throws Exception
+    {
+        // given
+        assertUU();
+
+        // then
+        lock.writerAndCleanerLock();
+        assertLL();
+
+        lock.writerAndCleanerUnlock();
+        assertUU();
+    }
+
+    private void assertThrow( Runnable unlock )
+    {
+        try
+        {
+            unlock.run();
+            fail( "Should have failed" );
+        }
+        catch ( IllegalStateException e )
+        {
+            // good
+        }
+    }
+
+    private void assertBlock( Runnable runLock, Runnable runUnlock ) throws ExecutionException, InterruptedException
+    {
+        Future<?> future = executor.submit( runLock );
+        shouldWait( future );
+        runUnlock.run();
+        future.get();
+    }
+
+    private void shouldWait( Future<?> future )throws InterruptedException, ExecutionException
+    {
+        try
+        {
+            future.get( 200, TimeUnit.MILLISECONDS );
```

Comment:
Hmm, that rule exists in neo4j-kernel. Could be a separate PR for moving it to something like neo4j-common and making use of it.